### PR TITLE
[SelectionDAG] Let ComputeKnownSignBits handle (shl (ext X), C)

### PR DESCRIPTION
**Pull request by bjope**
Add simple support for looking through ZEXT/ANYEXT/SEXT when doing ComputeKnownSignBits for SHL. This is valid for the case when all extended bits are shifted out, because then the number of sign bits can be found by analysing the EXT operand.

A future improvement could be to pass along the "shifted left by" information in the recursive calls to ComputeKnownSignBits. Allowing us to handle this more generically.

### CODE REVIEW

**Comment by bjope** [comment]
No regression tests yet. If anyone has a good idea how and where to put them, let me know.

The need for this was found downstream as DAG Combiner prefer `(SHL (ANY_EXTEND (ASHR ...)))` instead of `(SHL (SIGN_EXTEND (ASHR ...)))`, which actually is bad for ComputeNumSignBits.

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp [hunk 1]
```diff
@@ -4617,6 +4617,21 @@ unsigned SelectionDAG::ComputeNumSignBits(SDValue Op, const APInt &DemandedElts,
...
+      if (Op.getOperand(0).getOpcode() == ISD::ANY_EXTEND ||
+          Op.getOperand(0).getOpcode() == ISD::ZERO_EXTEND ||
+          Op.getOperand(0).getOpcode() == ISD::SIGN_EXTEND) {
+        SDValue Src = Op.getOperand(0);
+        EVT SrcVT = Src.getValueType();
+        SDValue ExtendedOp = Op.getOperand(0).getOperand(0);
+        EVT ExtendedOpVT = ExtendedOp.getValueType();
+        uint64_t ExtendedWidth =
+            SrcVT.getScalarSizeInBits() - ExtendedOpVT.getScalarSizeInBits();
+        if (ExtendedWidth <= *ShAmt) {
```
Comment (jayfoad) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:4629]: This seems wrong in the vector case, where the shift amounts could be different and *ShAmt is the max shift amount. In that case, some of the other shift amounts could be < ExtendedWidth.

Comment (bjope) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp]: Indeed it was wrong, thanks!
I've added a fixup to check for MimimumShiftAmount.

Comment (jayfoad) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp]: Looks better, thanks.

**Comment by bjope** [comment]
Added some test cases based on the `(sshlsat x, c) -> (shl x, c)` fold to verify number of sign bits that are computed.

File: llvm/test/CodeGen/X86/computenumsignbits-shl.ll [hunk 2]
```diff
@@ -0,0 +1,158 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc < %s -mtriple=x86_64-linux | FileCheck %s --check-prefix=X64
```
Comment (RKSimon) [inline llvm/test/CodeGen/X86/computenumsignbits-shl.ll:2]: (pedantic) rename to `known-signtbits-shl.ll` to

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp [hunk 3]
```diff
@@ -4615,12 +4615,30 @@ unsigned SelectionDAG::ComputeNumSignBits(SDValue Op, const APInt &DemandedElts,
       Tmp = std::min<uint64_t>(Tmp + *ShAmt, VTBits);
     return Tmp;
   case ISD::SHL:
-    if (std::optional<uint64_t> ShAmt =
+    if (std::optional<uint64_t> MaxShAmt =
             getValidMaximumShiftAmount(Op, DemandedElts, Depth + 1)) {
+      if (Op.getOperand(0).getOpcode() == ISD::ANY_EXTEND ||
+          Op.getOperand(0).getOpcode() == ISD::ZERO_EXTEND ||
+          Op.getOperand(0).getOpcode() == ISD::SIGN_EXTEND)
```
Comment (RKSimon) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:4622]: Use ISD::isExtOpcode(Op.getOperand(0).getOpcode())

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp [hunk 4]
```diff
@@ -4615,12 +4615,30 @@ unsigned SelectionDAG::ComputeNumSignBits(SDValue Op, const APInt &DemandedElts,
...
+        if (std::optional<uint64_t> MinShAmt =
+                getValidMinimumShiftAmount(Op, DemandedElts, Depth + 1)) {
+          SDValue Src = Op.getOperand(0);
+          EVT SrcVT = Src.getValueType();
+          SDValue ExtendedOp = Op.getOperand(0).getOperand(0);
+          EVT ExtendedOpVT = ExtendedOp.getValueType();
```
Comment (RKSimon) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:4628]: (pedantic) ExtendedOp isn't extended - it's the pre-extended source value

**Review by RKSimon — COMMENTED** [review]
A couple of minor comments

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp [hunk 5]
```diff
@@ -4615,12 +4615,28 @@ unsigned SelectionDAG::ComputeNumSignBits(SDValue Op, const APInt &DemandedElts,
...
+      uint64_t MaxShAmt = ShAmtRange->getUnsignedMax().getZExtValue();
+      uint64_t MinShAmt = ShAmtRange->getUnsignedMin().getZExtValue();
+      if (ISD::isExtOpcode(Op.getOperand(0).getOpcode())) {
+        SDValue Ext = Op.getOperand(0);
+        EVT ExtVT = Ext.getValueType();
+        SDValue Extendee = Ext.getOperand(0);
+        EVT ExtendeeVT = Extendee.getValueType();
+        uint64_t SizeDifference =
+            ExtVT.getScalarSizeInBits() - ExtendeeVT.getScalarSizeInBits();
+        if (SizeDifference <= MinShAmt) {
```
Comment (RKSimon) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:4629]: Add a explanation comment (similar to the patch summary)

**Review by RKSimon — APPROVED** [review]
LGTM - cheers

**Comment by bjope** [comment]
> LGTM - cheers

Thanks!

**Pull request merged**